hypothesis-python/src/hypothesis/internal/conjecture/engine.py

```diff
@@ -156,9 +161,21 @@ def __tree_is_exhausted(self):
 
     def __stoppable_test_function(self, data):
         """Run ``self._test_function``, but convert a ``StopTest`` exception
-        into a normal return.
+        into a normal return and avoid raising Flaky for RecursionErrors.
         """
+        depth = stack_depth_of_caller()
+        # Because we add to the recursion limit, to be good citizens we also add
+        # a check for unbounded recursion.  The default limit is 1000, so this can
+        # only ever trigger if something really strange is happening and it's hard
+        # to imagine an intentionally-deeply-recursive use of this code.
+        assert depth <= 1000, (
+            "Hypothesis would usually add %d to the stack depth of %d here, "
+            "but we are already much deeper than expected.  Aborting now, to "
+            "avoid extending the stack limit in an infinite loop..."
+            % (self.__recursion_limit, depth)
+        )
         try:
+            sys.setrecursionlimit(depth + self.__recursion_limit)
```

Comment:
Nit: You don't need to change this, but I wonder if it makes sense to set this instead to `depth + self.__recursion_limit - self.__base_stack_depth` where `self.__base_stack_depth = stack_depth_of_caller` in the `__init__`.

The reasoning basically being that we don't want to set the stack depth too unreasonably high (because if you set it super high you can get segfaults instead of `RecursionError` and there might be quite a lot of stack depth above us.

Reply:
This *is* kind of appealing, but in a way that makes me suspicious of my sense of taste... we're already asserting that the current stack depth is <= 1000, so I don't think the marginal benefit is quite worth making this code any hairier than it absolutely has to be.